### PR TITLE
batch of message transformation fixes

### PR DESCRIPTION
**Pull request by thalesmg**
Fixes https://emqx.atlassian.net/browse/EMQX-12683

Fixes https://emqx.atlassian.net/browse/EMQX-12684

Fixes https://emqx.atlassian.net/browse/EMQX-12678

Fixes https://emqx.atlassian.net/browse/EMQX-12682

Fixes https://emqx.atlassian.net/browse/EMQX-12679

Fixes https://emqx.atlassian.net/browse/EMQX-12677

Release version: e5.7.2

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_message_transformation/src/emqx_message_transformation.erl [hunk 1]
```diff
@@ -355,9 +358,11 @@ message_to_context(#message{} = Message, Payload, Transformation) ->
         node => node(),
         payload => Payload,
         peername => Peername,
+        pub_props => Props,
         publish_received_at => Message#message.timestamp,
         qos => Message#message.qos,
         retain => emqx_message:get_flag(retain, Message, false),
+        timestamp => Timestamp,
```
Comment (zmstone) [inline apps/emqx_message_transformation/src/emqx_message_transformation.erl:365]: Maybe it should just duplicate publish_received_at.
or if publish_received_at should be renamed to timestamp ?

File: apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl [hunk 2]
```diff
@@ -108,8 +108,7 @@ fields(transformation) ->
                 hoconsc:array(ref(operation)),
                 #{
                     desc => ?DESC("operation"),
-                    required => true,
-                    validator => fun validate_operations/1
+                    required => true
```
Comment (zmstone) [inline apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl:111]: ?
```suggestion
                    required => false
```

Comment (zmstone) [reply apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl:111]: and provided default value `[]`

File: apps/emqx_message_transformation/src/emqx_message_transformation.erl (hunk 1 above)
Comment (thalesmg) [reply apps/emqx_message_transformation/src/emqx_message_transformation.erl:365]: I was copying the semantics from rule engine: there, `timestamp` is just the time at which processing is happening, for whatever reason...
I also question the usefulness of this field, but Product asked for it: https://emqxio.slack.com/archives/C023T7W5N3Y/p1719909861660009?thread_ts=1718812121.788739&cid=C023T7W5N3Y

File: apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl (hunk 2 above)
Comment (thalesmg) [reply apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl]: I've added the default value and omitted `required`.  That should be equivalent and enough, I believe?

File: apps/emqx_message_transformation/src/emqx_message_transformation.erl (hunk 1 above)
Comment (thalesmg) [reply apps/emqx_message_transformation/src/emqx_message_transformation.erl:365]: Replied in DM with internal link to context.

**Review by ieQu1 — APPROVED** [review]

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3064)

**Pull request merged**